### ISSUE TITLE
Failure on Windows 10 v2004 and v2009 under HyperV with dynamic memory

### ISSUE THREAD

**Issue by WarrenArthur** [Velocidex/WinPmem#25]
**TLDR:**
The RC2 version of WinPMem v4 fails to run reliably on Windows 10 v2004 and v2009 (20H2) when in a HyperV guest with the "dynamic memory" option enabled. This option is unfortunately a HyperV default, but may be disabled. The failure occurs when the memory of the guest has been expanded at some point in the past due to pressure on memory within the guest. Prior to such memory expansion the memory dump operates as expected.

**Conditions tested:**
1. Execution under Windows 10 v2004 and v2009.
2. Operating system under HyperV with the "dynamic memory" option enabled.
3. Operating system under HyperV with the "dynamic memory" option disabled.
4. Operating system under VMware v16.1.
5. RAM allocations to virtual machines: 2GB, 4GB.
6. Processor allocations to virtual machines: 2, 4.
7. Memory pressure applied by running multiple instances of VisualStudio to consume memory prior to test.
8. Target output modes tested: both stdout and file.
9. All methods of memory acquisition (-0, -1, -2).

**Conditions NOT tested:**
1. Execution under any other Windows OS version.

**Failure condition noticed at:**
1. Windows 10 v2004 and v2009
2. Execution under HyperV with "dynamic memory" enabled.
3. Failure is independent of the number of cores and the RAM allocation.
4. Failure does NOT occur on VMware guests.
5. Failure does not occur when "dynamic memory" is disabled.

**Symptoms:**
1. Immediate error code received: -1 or -8.
2. Failure message:
"Failed to get memory geometry: The program issued a command but the command length is incorrect"

**Notes:**
1. Failure does not occur unless memory pressure has been applied to the guest at some point in the past since the guest has started.

**Comment by scudette** [comment]
Do we see any logs from the kernel? Can you please install debug viewer and capture kernel messages to help debug? 

Maybe in this configuration, the guest has many many physical memory ranges and they dont fix in the IOCTRL we sent to the kernel?

**Comment by WarrenArthur** [comment]
Thanks for the reply. I will allocate some time to follow-up on the information request and revert back.

**Comment by WarrenArthur** [comment]
As requested:
HyperV configuration: 2 Cores, 2GB Memory (dynamic memory option enabled)

![Winpmem kernel output](https://user-images.githubusercontent.com/50675611/105699421-00a71f00-5f10-11eb-95ae-808104e5c7f4.JPG)

**Comment by WarrenArthur** [comment]
HyperV configuration for the above.

![Winpmem issue, HyperV settings](https://user-images.githubusercontent.com/50675611/105699696-73b09580-5f10-11eb-8d85-8789b735e7dd.JPG)

Logs in text format:
[debug viewer -kernel - winpmem.LOG.txt](https://github.com/Velocidex/WinPmem/files/5866199/debug.viewer.-kernel.-.winpmem.LOG.txt)

**Comment by WarrenArthur** [comment]
For reference a similar output is attached for the case when "dynamic memory" is not used.
There is no failure in this situation, everything works as expected.

[winpmem screenshots static.zip](https://github.com/Velocidex/WinPmem/files/5866366/winpmem.screenshots.static.zip)

**Comment by vivianezw** [comment]
Guys, I wrote about that.
https://github.com/Velocidex/WinPmem/issues/10

Please uncheck the "Dynamic Memory" by setting your memory to a static size (and please tell if it removed the problem successfully). It's most likely not a Winpmem issue.
After a while, the memory ranges become **heavily** fragmented. You can check that with the Sysinternals Rammap tool from Microsoft. 
MS just don't seem to fix it ever.

**Comment by scudette** [comment]
Can we just increase the size of the buffer we send from user space to
accommodate all the ranges?

**Comment by WarrenArthur** [comment]
@scudette investigating any such solution would be appreciated from our side, since we have to run WinPMem on some machines where changing the dynamic memory option for HyperV is not available to us.

**Comment by vivianezw** [comment]
That was one idea I had in the past, but it's simply too much, I saw 8000 memory ranges one time!

Here is one: 
![memoryrangeFragmentationbug](https://user-images.githubusercontent.com/71179084/105989712-ce2d2b80-60a1-11eb-8969-7ed66abf8f06.png)

If it were only 100 or even 500, okay. But how far can this go? It could go over 10000 memory ranges to transfer. I mean, with fast I/O, sure, it's possible. But we do have static-sized Device I/O packets right now, that's also the reason why it fails (at least it fails in a controlled way...). 

OK, I'm right now reporting it (yes, right now), it would really be good if it was fixed.

**Comment by vivianezw** [comment]
Hi guys, I've tried multiple approaches using trial & error, and I've now been given an email to report the kernel bug. :-)
There's hope for this to be fixed.
